### PR TITLE
fix: unpin only the specifiers in the bundle a bun modules run executes

### LINKED ISSUE #11080: bug: server-side lockfile regeneration strips pinned import versions for transitively imported scripts (locks to "latest")
## Describe the bug

When the server asynchronously regenerates a Bun/TypeScript script's lockfile (dependency jobs — on script create/edit and on the relative-import relock cascade), a pinned import specifier such as `import * as wmill from "windmill-client@1.795.0"` is honored **only when it appears in the runnable's own entry file**. If the pin lives in a script the runnable imports (resolved as a module during the build), the module content is fetched through the `raw_unpinned` endpoint, which unconditionally strips `@version` from import specifiers (`remove_pinned_imports`). The lockfile generator then sees a bare `windmill-client`, defaults it to `"latest"` (`loader_builder.bun.js`), and `bun install` resolves npm's `latest` tag (currently 1.809.0). The regenerated lock is persisted over the deployed one (`UPDATE script SET lock = $1`).

Net effect: any runnable that reaches an npm dependency transitively gets its lock silently rewritten from the pinned version to `latest` within a minute or two of every deploy — even though the same import pinned directly in the entry file locks correctly, and even though `wmill generate-metadata` (CLI, local resolution) produces the correctly pinned lock for the same code.

We run v1.795.0 self-hosted and verified the causing code is unchanged in v1.809.0 (`bun_executor.rs` still wires the lock-gen loader to `raw_unpinned`; `remove_pinned_imports` and its call in the `raw_unpinned` handler are unchanged).

## To reproduce

Three minimal scripts (bun):

1. `pin_direct` — pin in its own file (locks correctly, the control):

```ts
import * as wmill from "windmill-client@1.795.0"
export function main() { return typeof wmill }
```

Save, wait for the lock job, check the lockfile: `"windmill-client": "1.795.0"` ✓

2. `pin_module` — same pinned import, exports a helper (also locks correctly):

```ts
import * as wmill from "windmill-client@1.795.0"
export function helper() { return typeof wmill }
```

3. `pin_consumer` — reaches the dependency only transitively:

```ts
import { helper } from "/u/<your_user>/pin_module"
export function main() { return helper() }
```

Create it after the first two exist (or edit it — see note), wait for the lock job, check the lockfile: `"windmill-client": "latest"` resolving `windmill-client@1.809.0` ✗

Actual behavior: step 3's lockfile declares `"latest"` despite the only import chain to the package carrying an explicit `@1.795.0` pin.

Note: if `pin_consumer` is created before `pin_module` finished deploying, its first lock job fails (`Failed to find relative import at .../raw_unpinned/...`) and is not retried — editing the consumer re-queues it, which then "succeeds" with `latest`.

Variant (how we hit it in production): edit a script that other scripts import, changing its lockfile hash — the `process_relative_imports` cascade re-queues lock jobs for every importer, flipping every importer whose npm dependency is transitive. A mass deploy that pins/changes many modules therefore flips many consumer locks a minute or two after the deploy converges.

## Expected behavior

Server-side lockfile regeneration preserves the exact version from the import specifier — `"windmill-client": "1.795.0"` — regardless of whether the pinned import lives in the entry file or in a transitively imported script. This is what the CLI already does for all cases since #10427 (shipped 1.795.0); server-side relock should behave the same, otherwise git-deployed locks and server-regenerated locks disagree forever for transitive imports.

## Screenshots

Not applicable — text lockfile diffs only:

```diff
   "dependencies": {
-    "windmill-client": "1.795.0"
+    "windmill-client": "latest"
   }
-    "windmill-client": ["windmill-client@1.795.0", "", {}, "sha512-Ay6v0lQ…"],
+    "windmill-client": ["windmill-client@1.809.0", "", {}, "sha512-Qbw6Bsm…"],
```

## Browser information

Not applicable — purely server-side (worker dependency jobs).

## Application version

`Running windmill version (backend): v1.795.0` (self-hosted; workers bundle bun 1.4.0 per Dockerfile). Verified the code responsible is byte-identical in v1.809.0 / `main`.

## Additional Context

Root-cause chain (v1.795.0 line numbers):

- `backend/windmill-worker/src/bun_executor.rs:935` — the relative-import loader used for lockfile generation fetches imported scripts from `raw_unpinned` (`.replace("RAW_GET_ENDPOINT", "raw_unpinned")`)
- `backend/windmill-api-scripts/src/scripts.rs:3350-3354` — the `raw_unpinned` handler applies `remove_pinned_imports(&content)` before serving (`unpin = true`)
- `backend/parsers/windmill-parser-ts/src/lib.rs:671-694` — `remove_pinned_imports()` strips `@version` from every import specifier (`IMPORTS_VERSION` regex)
- `backend/windmill-worker/loader_builder.bun.js:35-77` — dependencies are derived by scanning the **built chunk**; a specifier without a version defaults to `"latest"` (lines 52-55, 67-69) and `bun install` resolves it
- `backend/windmill-worker/src/worker_lockfiles.rs:471-486` — the regenerated lock is written over the deployed lock; `:522` (`process_relative_imports`) cascades relocks to importers when an imported script's lock hash changes

Supporting elimination evidence: a faithful local replica of `loader_builder.bun.js` on bun 1.4.0 produces `{"windmill-client":"1.795.0"}` for both the direct and the transitive case (module provided via plugin `onLoad({contents, loader:"tsx"})` to mirror the fetch path) — the builder and Bun preserve `pkg@version` when the input still has it. The only input that yields `"latest"` is module content whose pins were already stripped, i.e. exactly what `raw_unpinned` serves.

Impact: we manage all runnables as code (wmill sync + git) with CI-enforced pinned locks; every transitively importing runnable drifts to `latest` after each deploy, executing an untested version and wedging our sync automation.

Suggested fix: the `unpin` transformation has legitimate uses for execution/editor flows, but lockfile generation needs the pins. Simplest: have the lock-generation loader fetch the pinned variant of the raw endpoint (`unpin=false`) — a one-line change scoped to `bun_executor.rs:935`'s `RAW_GET_ENDPOINT` for the lockfile-generation build. Alternatively, derive dependencies from the pre-strip sources instead of the built chunk.

Related, same class of pin-loss during lockfile generation: #10425 and #10427 (CLI side, fixed in 1.795.0) — this is the server-side sibling that remains.

### PR DESCRIPTION
**Pull request by rubenfiszel**
## Summary

Follow-up to #11082. That PR fixed #11080 by restoring the loader's `RAW_GET_ENDPOINT` placeholder, so lock generation fetches imported scripts with their `pkg@version` pins.

One run path executes that lock-generation build directly. A Bun run with `modules` and no lock (the editor's Test button on a script with module tabs) uses `out/main.js` as its entry. That bundle keeps pinned specifiers, and bun resolves a pinned specifier outside `node_modules`, loading a second copy of the package from its global cache.

Since #11082 this covers imported scripts' pins as well as the entry's own. On a local instance, `import * as a from "is-number"` in the entry and `"is-number@6.0.0"` in an imported script gave `a !== b`.

The bundle's pins are now stripped before the run by a new `remove_pinned_import_specifiers`. It rewrites only the specifier literals of `import`/`export … from` and dynamic `import()`. The bundle holds the user's code, so other text is left as written, including a string equal to a pinned specifier. If the parser rejects the bundle, a warning is logged and the bundle runs as built, which is today's behavior.

**Nothing else changes.** `remove_pinned_imports` and `parse_expr_for_imports` are untouched (the `lib.rs` diff is additions only). Single-file runs, deployed runs, dedicated workers and the `raw_unpinned` endpoint take the same code paths as on `main`.

## Changes

- `backend/parsers/windmill-parser-ts/src/lib.rs`: new `remove_pinned_import_specifiers`. It collects specifier spans with a visitor and rewrites only those, skipping any span that does not sit exactly between a matching quote pair. `require()` is left alone because a script can shadow `require`.
- `backend/windmill-worker/src/bun_executor.rs`: apply it to `out/main.js` before it replaces `main.ts` for a modules run. On a parse error, log a warning and fall back to the unmodified bundle.
- `backend/parsers/windmill-parser-ts/tests/tests.rs`: unit test covering each specifier form. It also checks what must stay as written (a string, a comment, a shadowed `require`) and the offset edge cases (non-ASCII, BOM, CRLF).
- `backend/tests/bun_jobs.rs`: `test_bun_modules_run_loads_imported_pin_from_node_modules`. A modules run importing a script that pins `is-number@6.0.0` must load one instance and return the script's `"is-number@6.0.0"` string unchanged. It fails without the strip (`false`) and with `remove_pinned_imports` in its place (the string comes back as `"is-number"`).
- `backend/windmill-worker/loader.bun.js`: a comment at the endpoint placeholder recording why lock generation needs `raw`.
- `cli/test/sync_push_auto_metadata_repro.test.ts`: a comment named `raw_unpinned` as the endpoint lock generation uses. It is `raw`.

## Behavior change to note

When two imported scripts pin the same package at different versions, a multi-file preview used to load each version separately. It now uses the single version the lock resolved, which is the lowest. Deployed runs and single-file previews already did this (checked on a local instance: both return `["4.17.20", "4.17.20"]` for `lodash@4.17.20` and `lodash@4.17.21`). Multi-file previews now agree with them.

## Test plan

- [x] `cargo test -p windmill-parser-ts --test tests` (33) and `cargo test --features quickjs --test bun_jobs --test script_modules -- --test-threads=10` on an empty `WINDMILL_DIR` (35 + 2, including #11082's test).
- [x] Property check of the new function over 57,946 real JS/TS files (the repo plus `frontend/` and `cli/` `node_modules`):
  - It never fails where `remove_pinned_imports` parses.
  - Files without pins come back byte-identical, and its output matches `remove_pinned_imports` on every one of them.
  - With every bare import specifier pinned (11,864 files), bun's own `scanImports` finds no pinned import left anywhere. Remaining pins sit only in comments, strings and TypeScript type positions.
  - All of this also holds after CRLF conversion and with a leading BOM.
  - Slowest file: 10 ms.
- [x] 3,000 bundles built the way lock generation builds them (`external: ["*"]`) from pinned corpus files. All parse, so the fallback never triggers. None loads a pinned import after unpinning, and every edit is a pure deletion of `@version` text (6,319 pins removed).
- [x] An 18-case matrix of multi-file previews on a local instance, run with and without the strip. Cases cover pins in the entry, in module files and in imported scripts; conflicting versions; dynamic `import()`; `require()`; re-exports with subpaths; scoped packages; type-only imports; BOM and CRLF; Unicode; `windmill-client`; modern syntax; `//nodejs`; `//nobundling`; and a string equal to a specifier.
  - No case that works without the strip breaks with it.
  - Five go from two package instances to one.
  - The conflicting-version case changes as described above.
  - One case (a module file importing a workspace script without `.ts`) fails identically in both, in bun's build step, before the strip runs.
- [ ] Dedicated-worker Bun runs (EE-gated tests, not run locally). They do not reach the changed path.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01WCq4nkUjiZBnPPMuGYCo6w

### CODE REVIEW

File: backend/parsers/windmill-parser-ts/src/lib.rs [hunk 1]
```diff
@@ -673,24 +677,92 @@ lazy_static::lazy_static! {
...
+        FileName::Custom("main.d.ts".into()).into(),
+        code.to_string(),
+    );
+    let module = parse_module_source(&fm)?;
+    let mut specifiers = ImportSpecifierSpans(vec![]);
+    specifiers.visit_module(&module);
+    specifiers.0.sort_by_key(|s| s.lo);
+
+    // Spans index the parsed source, which the source map stripped of any UTF-8 BOM.
+    let bom = code.len() - fm.src.len();
```
[the remark this answers is not in the document]
Comment (rubenfiszel) [reply backend/parsers/windmill-parser-ts/src/lib.rs]: Addressed in 72f3c17. `bom` is now derived from a leading U+FEFF in `code`, not from the length difference, so it is always 0 or 3. Every span must also land on a matching quote pair, or it is left as written. The unit test has a two-specifier CRLF case, and it passes, so swc 14 does not normalize line endings here (its `new_source_file` only calls `remove_bom`).

**Comment by rubenfiszel** [comment]
Re the round-3 Claude nit (an unshadowed `require("pkg@1.2.3")` keeps its pin): I'm leaving this as is.

- **Nothing regresses except a side effect.** On `main`, `remove_pinned_imports` only ever targeted static import/export specifiers. A pin that appears only in a `require()` was never stripped. The one narrowed case is a script that both statically imports and `require`s the same pinned specifier, where the old global `str::replace` happened to rewrite the `require` too.
- **Scope.** Rewriting `require` safely needs binding analysis, such as the suggested file-wide pre-pass for any `require`/`__require` binding. That is more machinery in a helper this PR has already had to narrow once, for a pattern that is rare in Bun TypeScript scripts.
- **The guard already exists.** The shadowed-`require` case in `test_remove_pinned_imports_rewrites_only_specifiers` fails if `require` handling is re-added without thinking about shadowing. An extra assertion that an unshadowed `require` keeps its pin would lock in a limitation and block a correct scope-aware version later, so I haven't added one.

**Comment by rubenfiszel** [comment]
✅ Review round clean @ 9cbf8022ca046e40c98dd1a701ca4fbb6142683d
